### PR TITLE
react-router 6 upgrade

### LINKED ISSUE #822: [Tech Debt] Upgrade `react-router` to v6
Upgrade `react-router` to v6. There is already some work started here, but not finished. Search for `react-router-dom-v5-compat`

### PR DESCRIPTION
**Pull request by trevdor**
Nearly-complete upgrade to v6 conventions, following [this guide](https://reactrouter.com/en/main/upgrading/v5#upgrade-to-react-router-v6).

Draft because I'm struggling to finish (short on time and not sure I understand the full purpose of the particular way `history` is manipulated in Actual). 
But this upgrade feels important to complete before I move to mobile transaction entry, which will likely bring a new set of routes on top of what we have.

### Remaining Work
- [x] Remove all direct usage of the `history` package. (3 spots left)
   * 2 `createBrowserHistory` and 1 `createLocation`
   * tricky bit: our `patchedHistory`, made available globally `window` (search `window.__history` to find usages)
   * Is `RouterProvider` the RR6 replacement for this? But will it do all we were doing with our global patched history? (See Note below)
- [x] Find a replacement type for `RouteComponentProps` in [common.tsx](https://github.com/actualbudget/actual/pull/1066/files#diff-db00a4331c3d9b4705f5e89a2719873651418d5118565e5084ad4341c984b422R17)
- [x] Regression test ButtonLink since it had its `withRouter` removed

**Bugs**
- [x] Transaction data not refreshing as you transition between accounts
- [ ] Mobile: individual account ledgers not loading (`/account/:id`)

### Note
_mis-remembered. Disregard_
~~Our global undo feature appears meant to take you back to the route you were at for the undone step. (e.g. you change a transaction in `/account/<uuid>`, visit the Budget page, then press CMD+Z -- you should jump back to `/account/<uuid>` -- honestly, I suspect we've broken this over the past several months, but haven't pinpointed when)~~ 

Closes #822

### CODE REVIEW

File: packages/desktop-client/src/components/FinancesApp.js [hunk 1]
```diff
@@ -62,78 +61,100 @@ function NarrowNotSupported({ children, redirectTo = '/budget' }) {
   return isNarrowWidth ? <Redirect to={redirectTo} /> : children;
 }
 
-function Routes({ location }) {
```
Comment (trevdor) [inline packages/desktop-client/src/components/FinancesApp.js:65]: `Routes` is a RR v6 component

File: packages/desktop-client/src/components/FinancesApp.js [hunk 2]
```diff
@@ -170,14 +191,13 @@ function NavTab({ icon: TabIcon, name, path }) {
...
-      style={{
+      style={({ isActive }) => ({
         alignItems: 'center',
-        color: '#8E8E8F',
+        color: isActive ? colors.p5 : '#8E8E8F',
         display: 'flex',
         flexDirection: 'column',
         textDecoration: 'none',
-      }}
-      activeStyle={{ color: colors.p5 }}
```
Comment (trevdor) [inline packages/desktop-client/src/components/FinancesApp.js:180]: `activeStyle` prop has been removed from NavLink in v6

File: packages/desktop-client/src/components/FinancesApp.js [hunk 3]
```diff
@@ -284,59 +304,51 @@ function FinancesApp(props) {
 
   return (
     <Router history={patchedHistory}>
-      <CompatRouter>
```
Comment (trevdor) [inline packages/desktop-client/src/components/FinancesApp.js:287]: Remove the compat router -- the rest is indentation here until the <Route> block.

File: packages/desktop-client/src/components/sidebar.js [hunk 4]
```diff
@@ -183,7 +170,6 @@ function SecondaryItem({
       <ItemContent
         style={linkStyle}
         to={to}
-        exact={exact}
```
Comment (trevdor) [inline packages/desktop-client/src/components/sidebar.js:186]: v6 routes do not support `exact`

File: packages/desktop-client/src/components/common.tsx [hunk 5]
```diff
@@ -155,25 +155,22 @@ export function Link({ style, children, ...nativeProps }: LinkProps) {
...
 
 export function AnchorLink({
   to,
-  exact,
   style,
   activeStyle,
   children,
 }: AnchorLinkProps) {
-  let match = useRouteMatch({ path: to, exact: true });
+  let match = useMatch({ path: to });
```
[the remark this answers is not in the document]
Comment (trevdor) [reply packages/desktop-client/src/components/common.tsx:169]: The matching algo changed to give [the most specific match](https://reactrouter.com/en/main/start/overview#ranked-route-matching), such that `exact` is rarely needed. 

https://reactrouter.com/en/main/upgrading/v5#relative-routes-and-links
> `<Route exact>` is gone. Instead, routes with descendant routes (defined in other components) use a trailing * in their path to indicate they match deeply

So if we need exact, we'd add that trailing `*` wherever the `to` is passed from.

**Comment by j-f1** [comment]
> Our global undo feature appears meant to take you back to the route you were at for the undone step. (e.g. you change a transaction in /account/<uuid>, visit the Budget page, then press CMD+Z -- you should jump back to /account/<uuid> -- honestly, I suspect we've broken this over the past several months, but haven't pinpointed when)

looked at the initial (open source) commit and it doesn’t seem to work there either

**Comment by trevdor** [comment]
> > Our global undo feature appears meant to take you back to the route you were at for the undone step. (e.g. you change a transaction in /account/, visit the Budget page, then press CMD+Z -- you should jump back to /account/ -- honestly, I suspect we've broken this over the past several months, but haven't pinpointed when)
> 
> looked at the initial (open source) commit and it doesn’t seem to work there either

Thanks for checking, @j-f1! I had gone back several versions but never confirmed how it was early on. Must have remembered wrong.

**Review by MatissJanis — APPROVED** [review]
Amazing

**Comment by trevdor** [comment]
@j-f1 Probably worth merging as you did so other work can build on it, but there’s a major bug left on this one. I’ll open a new issue where we can get it fixed before the next release.

**Comment by j-f1** [comment]
oop!

**Pull request merged**